### PR TITLE
make Roc implementation way simpler and more idiomatic

### PR DESCRIPTION
**Pull request by bhansconnect**
The first version was my personal version that was messing around with Data Oriented Design and other hacks.
This version is way cleaner, easier to follow, and more idiomatic.
Should be much closer, to other functional languages.

### CODE REVIEW

**Review by bhansconnect — COMMENTED** [review]
The only struggle with the CodeOwners setup is that I can't approve my own pull request even if it only touches roc code.

Anyway, this is tested, just updates Roc code, and gets rid of some overly complex ways of doing things. Instead it is much simpler and should be more readable to most people.

**Comment by ThePrimeagen** [comment]
will be merging :)

**Review by ThePrimeagen — APPROVED** [review]

**Comment by ThePrimeagen** [comment]
real talk, i have NO idea what this does

@bhansconnect i am going to make you code owner and i want you to review any more roc prs :)

**Pull request merged**
